### ISSUE TITLE
createSemanticDiagnosticsBuilderProgram is confusing and error prone

### ISSUE THREAD

**Issue by russelldavis** [microsoft/TypeScript#32385]
This is a follow up to #29176 which was never resolved. To clear up some confusion I think existed in the conversation there, here's what I think the crux of the issue is:

In the example at https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#writing-an-incremental-program-watcher, the code uses `createSemanticDiagnosticsBuilderProgram`. It is strongly implied by the comments there (and the name of the function itself, in contrast with `createEmitAnd...`) that the example will *not* emit code. But it actually does, for reasons explained at https://github.com/Microsoft/TypeScript/issues/29176#issuecomment-450232791. 

Not only does it emit code, but it does so several times slower than optimal, since it is not using the optimized `EmitAnd...` builder.

#29176 explains how it all works, but the slow & confusing example code remains. So, what do you think about the following?

1) Fix the example. Two options for that: (a) Change it to use `createEmitAndSemanticDiagnosticsBuilderProgram`. (b) Change it ~~to alter `afterCreateProgram`~~ so that it does not emit (EDIT: an easy way to do this without touching `afterCreateProgram` is to just set `{noEmit: true}` as the second arg to `createWatchCompilerHost`). Or create an example for both.
2) Throw an exception (or at least log a warning?) when calling `emit` with a non-emit builder (`SemanticDiagnosticsBuilderProgram`), since doing so will be much slower, and is probably a mistake. Or are there valid use cases for doing that?

Exploring that last point even further, it's unclear why the non-emit builder needs to exist at all. The only thing it seems to optimize is avoiding appending some filenames to an array (`affectedFilesPendingEmit`), which doesn't seem consquential. I may be missing something, but I wonder if this could all be simplified by unifying the two kinds of `BuilderProgram` and simply avoiding a call to `emit` when you don't want emitting (which you have to do anyway, even when using the non-emit builder).

Thanks!

cc @sheetalkamat @0xorial @weswigham (from #29176)

**Comment by sheetalkamat** [comment]
cc @DanielRosenwasser 
There exists two builders because lot of times, user of semantic diagnostics builder knows what files it wants to emit (eg. loaders for webpack). So this provides the way to be able to get diagnostics in optimal fashion in those scenarios. As explained earlier these users use emit with already known set of files it wants to emit so it isn't really slow as it is similar to emit and semantic diagnostics builder without builder having to track which files are emitted and which are not and that's the main distinction between the two builders where the emit and semantic builder, marks some file's change application complete only when that file (and related affected files) are emitted. It also needs to emit it in specific order since that's how the tracking works (it doesn't calculate all files to emit and that work is unnecessary when someone knows what files to emit and its their own order they are going to call in with.

**Comment by russelldavis** [comment]
@sheetalkamat thanks, that helps clarify. Given that, I think the issue is with naming, documentation, and the sample code.

It's now clear that `SemanticDiagnosticsBuilderProgram` **is** intended to be used with emit, so hopefully you can see why the other builder having the same name prefixed by `EmitAnd` is misleading. The names also don't hint at the difference in emitting all files vs affected files only (hence the confusion about performance issues).

Given the actual behavior differences, I'd propose naming them something like `AffectedFilesBuilderProgram` and `AllFilesBuilderProgram`, or something similar that captures the important difference between them.

The comments in the example code could also use some clarification:

> // Between `createEmitAndSemanticDiagnosticsBuilderProgram` and
  // `createSemanticDiagnosticsBuilderProgram`, the only difference is emit.
  // For pure type-checking scenarios, or when another tool/process handles emit,
  // using `createSemanticDiagnosticsBuilderProgram` may be more desirable.

"The only difference is emit" should say something like "the difference is that the latter will process *all* files, while the former will only process changed files and their transitive dependencies".

And finally, I think it makes sense for the example itself to be updated to use the builder that only processes affected files. If you're using a watcher, you most likely only want to process the files that changed.

Do those changes make sense?

**Comment by sheetalkamat** [comment]
I don't think we intend to change the names for the builder since these are there for long time. But adding @DanielRosenwasser to see if there is some clarification that needs to be done in samples.

**Comment by DanielRosenwasser** [comment]
Hey, thanks for looking into this @russelldavis - I think that comment suggestion is a change we'd be happy to accept on https://github.com/Microsoft/TypeScript-wiki/.

I'd personally be okay with an API rename provided we could deprecate the original names, but if this would be too confusing for existing users, maybe providing this directly in the JSDoc for each function and type would be good enough. That's also a PR we would be happy to take!

**Comment by Bobris** [comment]
I was also very surprised when host.writeFile was called for SemanticDiagnosticsBuilderProgram. But what I also found that watching builder is nearly twice slower compared to "normal" build without emit.
Note I had to monkey patch isSourceFileFromExternalLibrary to always return false, so incremental rechecks are relatively fast (it is monorepo with yarn workspaces, and no subprojects). And TypeScript is hosted inside ChakraCore. I know very corner case, so just for your interest.
